File: this-is-coding-test/14장_실패율.js
const solution = (N, stages) => {
  // 오름차순 정렬
  stages.sort((a, b) => a - b);

  // 사용자 수
  let players = stages.length;
  // 동일 스테이지 사용자 수
  let stageCnt = 0;
  // stages 를 for loop 할 시작 인덱스
  let startIdx = 0;
  // 실패율
  let failRate = [];

  // 1 부터 N 까지
  // N + 1 은 구해야할 실패율 범위에 포함되지 않아서 확인하지 않아도 된다
  for (let stage = 1; stage <= N; stage++) {
    // startIdx 부터 stages 길이 - 1 만큼 for loop
    for (let idx = startIdx; idx < stages.length; idx++) {
      if (stage === stages[idx]) {
        stageCnt += 1;
        continue;
      }

      startIdx = idx;
      break;
    }
    
    // 안쪽 for 문을 끝까지 돌고 그대로 끝나버린 경우가 있어서
    // failRate, players, stageCnt 처리를 
    // break 문 위에서 하지 않고 여기서 수행
    failRate.push([stageCnt / players, stage]);
    players -= stageCnt;
    stageCnt = 0;
  }

  return failRate.sort((a, b) => a[0] === b[0] ? a[1] - b[1] : b[0] - a[0]).map(v => v[1]);
}

console.log(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]));
console.log(solution(4, [4, 4, 4, 4, 4]));
console.log(solution(3, [4, 1, 1, 1, 4]));

File: this-is-coding-test/14장_실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        HashMap<Integer, Integer> users = new HashMap<Integer, Integer>();

        for (int stage: stages) {
            if (users.containsKey(stage)) {
                Integer cnt = users.get(stage);
                users.put(stage, cnt + 1);
            } else {
                users.put(stage, 1);
            }
        }

        HashMap<Integer, Double> failRate = new HashMap<Integer, Double>();;
        int total = stages.length;

        for (int i = 1; i <= N; i++) {
            int userCnt = 0;

            if (users.containsKey(i)) {
                userCnt = users.get(i);
            }

            // 0 으로 나누지 않도록 예외 처리
            if (total == 0) {
                failRate.put(i, 0.0);
            } else {
                failRate.put(i, (double) userCnt / total);
            }

            total -= userCnt;
        }

        ArrayList<Integer> stageList = new ArrayList<>(failRate.keySet());
        // Collections.sort(stageList);
        // stageList 를 stage 번호 순서로 오름차순 정렬하지 않아도 오답이 발생하지 않는다
        // keySet 메소드는 map 의 키 순서를 보장하지 않는데
        // 어떻게 실패율이 같을 때 stage 번호 순서대로 정렬이 잘 됐는지 아직 이해하지 못했다
        stageList.sort((s1, s2) -> Double.compare(failRate.get(s2), failRate.get(s1)));
        return stageList.stream().mapToInt(Integer::intValue).toArray();
    }
}

// 참고
// https://bada744.tistory.com/54
// https://codechacha.com/ko/java-sort-map/
// https://velog.io/@jwkim/java-arraylist-array-type-conversion
// https://school.programmers.co.kr/questions/50861
// https://recordsoflife.tistory.com/1263
// https://stackoverflow.com/questions/43871017/are-the-hashmap-entries-always-sorted-by-key-if-the-key-is-of-type-integer
